### ISSUE TITLE
VisualQC Outlier Detection Error

### ISSUE THREAD

**Issue by Leema-Krishna-Murali** [raamana/visualqc#52]
Hi @raamana,

I was testing the outlier detection module in VisualQC (latest version) and I see the same error as above. Please can assist in how to resolve this error:

$ visualqc_freesurfer --fs_dir '/mnt/fsx/Screening_3DT1w_FSResults' -i '/home/ubuntu/Desktop/visualqc_sub_list.txt' -olt subcortical

visualqc version 0.5.1+25.g1d9d10c for Freesurfer QC
	Time stamp : 2022-03-02 20:18:52

10 subjects are usable for review.
Input folder: /mnt/fsx/fsx/nbg/e2609/g000_301_Screening_3DT1w_Recommend_FSResults_NEW
Output folder: /mnt/fsx/fsx/nbg/e2609/g000_301_Screening_3DT1w_Recommend_FSResults_NEW/visualqc
Preprocessing data - please wait .. 
	(or contemplate the vastness of universe! )
Extracting feature type: subcortical
Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/visualqc-0.5.1+25.g1d9d10c-py3.8.egg/visualqc/workflows.py", line 368, in extract_features
    self.feature_paths[feat_type] = self.feature_extractor(self, feat_type)
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/visualqc-0.5.1+25.g1d9d10c-py3.8.egg/visualqc/readers.py", line 176, in gather_freesurfer_data
    features = np.vstack([read_aseg_stats(qcw.fs_dir, id) for id in qcw.id_list])
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/visualqc-0.5.1+25.g1d9d10c-py3.8.egg/visualqc/readers.py", line 176, in <listcomp>
    features = np.vstack([read_aseg_stats(qcw.fs_dir, id) for id in qcw.id_list])
**AttributeError: 'FreesurferRatingWorkflow' object has no attribute 'fs_dir'
Unable to extract subcortical features! skipping..
Features required for outlier detection are not available - skipping it.**
Attempting to generate the surface visualizations of parcellation ...

Thank you,
Leema

_Originally posted by @Leema-Krishna-Murali in https://github.com/raamana/visualqc/issues/23#issuecomment-1057353637_

**Comment by raamana** [comment]
Thanks Leema for reporting this - it seems like a bug, let me look into the code and get back to you shortly.

**Comment by raamana** [comment]
I just pushed a 0.5.2 release - could you update with the following commands and retry? thanks.

```bash
pip uninstall visualqc
pip install -U visualqc --no-cache-dir
```

**Comment by raamana** [comment]
I must mention the outlier detection methods implemented are rather very simple, and haven't been extensively validated to answer any serious questions.

for the type you choose, it will use Isolation Forest algorithm to look for any subjects with very small or very large subcortical volumes (dataset size matters, so try use more than 10 subjects if you can). If that isn't the case you notice, I would love to hear from you on your feedback etc

**Comment by Leema-Krishna-Murali** [comment]
Thank you @raamana for taking a look at the issue. Really appreciate your time and effort.

Previously, I did check the source code on the outlier detection (outlier.py)  having Isolation Forest algorithm. Just wanted to test it out along with the GUI that VisualQC provides. And then later try out other outlier algorithms for my dataset. 

I did try those commands you suggested and used a dataset of 10 subjects. I got the following error:

Preprocessing data - please wait .. 
	(or contemplate the vastness of universe! )
Extracting feature type: subcortical
**Assitance with outlier detection did not succeed.**
Proceeding by disabling it. Stack trace below:

Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/visualqc/workflows.py", line 401, in detect_outliers
    features = gather_data(self.feature_paths[feature_type],
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/visualqc/readers.py", line 196, in gather_data
    features = np.vstack([np.genfromtxt(path_list[sid]) for sid in id_list])
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/visualqc/readers.py", line 196, in <listcomp>
    features = np.vstack([np.genfromtxt(path_list[sid]) for sid in id_list])
IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/visualqc/workflows.py", line 404, in detect_outliers
    raise IOError('Unable to read/assemble features for outlier '
**OSError: Unable to read/assemble features for outlier detection. Skipping them!**

Thanks again,
Leema

**Comment by raamana** [comment]
can you share the versions of python, numpy and scipy you are using? seems like some syntax issue.

**Comment by raamana** [comment]
I've identified and fixed another bug, just pushed a 0.5.3 release - could you update with the following commands and retry? thanks.

```bash
pip uninstall visualqc
pip install -U visualqc --no-cache-dir
```

**Comment by raamana** [comment]
if this fails again, there is a lot of version info printed at the top of the output log, remember to share that here.

**Comment by raamana** [comment]
closing this now! also FYI latest release 0.6 adds many improvements and fixes, from #57
